statsmodels/doe/latin.py

```diff
@@ -0,0 +1,200 @@
+    ----------
+    [1] Mckay et al., "A Comparison of Three Methods for Selecting Values of
+    Input Variables in the Analysis of Output from a Computer Code",
+    Technometrics, 1979.
+
+    """
+    if centered:
+        r = 0.5
+    else:
+        r = np.random.random_sample((n_sample, dim))
+
+    q = np.random.random_integers(low=1, high=n_sample, size=(n_sample, dim))
+
+    sample = 1. / n_sample * (q - r)
+
+    # Sample scaling from unit hypercube to feature range
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = sample * (max_ - min_) + min_
+
+    return sample
+
+
+def optimal_design(dim, n_sample, bounds=None, start_design=None, niter=1,
+                   force=False):
+    """Optimal design.
+
+    Optimize the design by doing random permutations to lower the centered
+    discrepancy.
+
+    Centered discrepancy based design show better space filling robustness
+    toward 2D and 3D subprojections. Distance based design better space filling
+    but less robust to subprojections.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    start_design : array_like (n_samples, k_vars)
+        Initial design of experiment to optimize.
+
+    Returns
+    -------
+    sample : array_like (n_samples, k_vars)
+        Optimal Latin hypercube Sampling.
+
+    References
+    ----------
+    [1] Damblin et al., "Numerical studies of space filling designs:
+    optimization of Latin Hypercube Samples and subprojection properties",
+    Journal of Simulation, 2013.
+
+    """
+    doe = start_design
+    if doe is None:
+        doe = orthogonal_latin_hypercube(dim, n_sample, bounds)
+
+    def _perturb_doe(x, sample, bounds):
+        """Disturbe the Design of Experiment.
+
+        Parameters
+        ----------
+        x : list of int
+            It is a list of:
+                idx : int
+                    Index value of the components to compute
+        sample : array_like (n_samples, k_vars)
+            Sample to perturb.
+        bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+            Desired range of transformed data. The transformation apply the
+            bounds on the sample and not the theoretical space, unit cube. Thus
+            min and max values of the sample will coincide with the bounds.
+
+        Returns
+        -------
+        discrepancy : float
+            Centered discrepancy.
+
+        """
+        col, row_1, row_2 = map(int, x)
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
```

Comment:
this is most likely a bug.
I never checked, but numpy does immediate inplace replacement
(IIUC there was initially a numpy random shuffle or permutation bug because of this)

Reply:
Not sure I understand. You mean this code is not working? I should be using numpy permutation instead?

Reply:
What I thought, but maybe it is not relevant if doe[row_2, col], doe[row_1, col] are scalars

In general, you need to do this in two steps and store a copy of the second value in a temp variable when using numpy arrays. However, it seems to work for scalars. (I'm using numpy 1.11.2)

In the first case xx[i] is mutable and changed before the second assignment, In the second case it works, I guess because it uses scalars.

```
>>> xx = np.arange(5)[:,None]
>>> i, j = 3, 4; xx[i], xx[j] = xx[j], xx[i]
>>> xx  
array([[0],
       [1],
       [2],
       [4],
       [4]])

>>> xx = np.arange(5)[:,None]
>>> i, j = 3, 4; xx[i, 0], xx[j, 0] = xx[j, 0], xx[i, 0]
>>> xx
array([[0],
       [1],
       [2],
       [4],
       [3]]
``